### PR TITLE
part 1.9, 'Function Types'

### PR DESCRIPTION
**Pull request by Zelenya**
Didn't add last code samples, no idea what to do with these.

```haskell
absurd (Void a) = absurd a
```
```haskell
newtype Void = Void Void
```

### CODE REVIEW

**Comment by SystemFw** [comment]
```scala
  // ignoring the newtype vs data distinction
  case class Void(v: Void)

  // both valid translations

  // this is more faithful
  def absurd[A]: Void => A = _ match {
    case Void(v) => absurd(v)
  }

  // this is more idiomatic
  def absurd[A](v: Void): A = absurd(v.v)

```

File: src/main/tut/1.9-function-types.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
+```Haskell
+eval :: ((a⇒b) × a) -> b
+```
+```scala
+def eval[`A=>B`, A, B]: ((`A=>B`, A)) => B
```
Comment (SystemFw) [inline src/main/tut/1.9-function-types.md]: This is incorrect, the following compiles:
```scala
  def eval[`A=>B`, A, B]: ((`A=>B`, A)) => B = ???
  def no = eval[String, Boolean, Double]
```
The literal translation of your code should be
```scala
def eval[`=>`[_, _], A, B]
```
but that's not what the text is saying. `⇒` and `×` are not type parameters there, but concrete types. The text says that `eval` is a morphism associated with the function object, whereas if you add it as a type parameter it means that it'll work `forall` the types of the given shape.

File: src/main/tut/1.9-function-types.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+```
+................
+```Haskell
+g :: z × a -> b
+```
+```scala
+def g[Z, A, B]: ((Z, A)) => B
```
Comment (SystemFw) [inline src/main/tut/1.9-function-types.md]: See the first comment re:  `×`

File: src/main/tut/1.9-function-types.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+```
+................
+```Haskell
+h :: z -> (a⇒b)
+```
+```scala
+def h[Z, `A=>B`]: Z => `A=>B`
```
Comment (SystemFw) [inline src/main/tut/1.9-function-types.md]: See the first comment re: `⇒`

File: src/main/tut/1.9-function-types.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+A => B => C
+```
+................
+```Haskell
+catstr :: String -> String -> String
+catstr s s’ = s ++ s’
+```
+```tut:silent
+val catstr: (String, String) => String = {
+  case (s, sᛌ) => s ++ sᛌ
```
Comment (SystemFw) [inline src/main/tut/1.9-function-types.md]: the case is not necessary here, you're taking two arguments, not a pair (in scala these two concepts aren't the same):
```scala
  val catstr: (String, String) => String = (s, sᛌ) => s ++ sᛌ
```

That said, the book is talking specifically about currying here, so this should probably be curried `val catstr: String => String => String`

I'd also recommend picking something less confusing then `sᛌ`. It' easy to confuse it with `s'` which won't compile in Scala

**Comment by Zelenya** [comment]
@SystemFw Thanks. Will try to figure it out.

There is also one _absurd_ at the end. And I borrowed it from previous chapters. Is there a way to combine it with what you suggested for other snippets?

```haskell
absurd :: Void -> a
```
```scala
 def absurd[A]: Nothing => A
 ```

File: src/main/tut/1.9-function-types.md (hunk 4 above)
Comment (Zelenya) [reply src/main/tut/1.9-function-types.md]: Followed @kailuowang suggestion for prime symbol in readme. Should it be removed from code and readme?

**Comment by SystemFw** [comment]
@Zelenya 
arguably 

```scala
def absurd[A]: Nothing => A
```

Is the correct translation in scala, since scala already has a bottom type. This won't allow you to define it though (haskell has to define it explicitly with that recursive newtype). I'd go with the `Nothing` version

File: src/main/tut/1.9-function-types.md (hunk 4 above)
Comment (SystemFw) [reply src/main/tut/1.9-function-types.md]: oops, my bad then :) Leave it as is

**Comment by Zelenya** [comment]
@SystemFw Updated

File: src/main/tut/1.9-function-types.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
+```Haskell
+eval :: ((a⇒b) × a) -> b
+```
+```scala
+def eval[A, B]: (`A⇒B` × A) => B
```
Comment (SystemFw) [inline src/main/tut/1.9-function-types.md]: it should be:
```scala
(A `⇒` B)
```
not: 
```scala
`A⇒B`
```
The latter case means a type of kind `*` whose name is `A⇒B`, the former a type of kind `* -> * ->` named `⇒` an parameterised by `A` and `B`. You also need to add the parentheses I _think_

File: src/main/tut/1.9-function-types.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+```
+```scala
+def g[Z, A, B]: (Z × A) => B
+```
+................
+```Haskell
+h :: z -> (a⇒b)
+```
+```scala
+def h[Z]: Z => `A⇒B`
```
Comment (SystemFw) [inline src/main/tut/1.9-function-types.md]: Same as above

File: src/main/tut/1.9-function-types.md (hunk 5 above)
Comment (Zelenya) [reply src/main/tut/1.9-function-types.md]: Yeah, that is what I had before, but in the book the object name is "a⇒b", and it didn't fell right to have  (A `A⇒B` B) or change the name. Anyways, it is much easier to read, so I will revert it back.

**Comment by kailuowang** [comment]
:+1:

**Pull request merged**